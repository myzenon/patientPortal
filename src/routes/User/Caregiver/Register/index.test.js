import React from 'react'
import { CaregiverRegister } from './index'
import ReactGA from 'react-ga';

   // This should be part of your setup
   ReactGA.initialize('UA-121783543-1', {
       'debug': true
     });;

jest.mock("react-ga")

jest.mock('assets/policy/nopp.html', () => '<meta charset="UTF-8">\n<meta version="v1.0">\nnopp content')
jest.mock('assets/policy/eula.html', () => '<meta charset="UTF-8">\n<meta version="v1.0">\neula content')
jest.mock('assets/policy/terms.html', () => '<meta charset="UTF-8">\n<meta version="v1.0">\nterms content')
jest.mock('assets/policy/access.html', () => '<meta charset="UTF-8">\n<meta version="v1.0">\naccess content')

const propsFromRedux = {
    page: {
        apiData: {
            passwordPattern: /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#\$%\^&\*\=])(?=.{8,})/,
            passwordMsg: 'Password must be at least 8 characters long and contain at least one character, one number and one special characters including !@#$%*=',
            passwordCriteria: [
                '8 characters',
                'One uppercase letter',
                'One lowercase letter',
                'One number',
                'One special character !@#$%^&*='
            ]
        }
    },
    callSyncAPIGateway: () => Promise.resolve()
}

test('renders correctly', done => {
    const props = {
        ...propsFromRedux,
    }
    const wrapper = mount (
        <CaregiverRegister {...props} />
    )
    setTimeout(() => {
        expect(toJson(wrapper.render())).toMatchSnapshot()
        done()
    }, 0)
})

// test('toggle switch showing password | show password between as clear text and as *******', done => {
//     const props = {
//         ...propsFromRedux,
//     }
//     const wrapper = mount (
//         <CaregiverRegister {...props} />
//     )
//     setTimeout(() => {
//         wrapper.find('span[className="switch switchOff"]').simulate('click')
//         expect(wrapper.find('input[name="password"]').props().type).toBe('text')
//         expect(wrapper.find('input[name="rePassword"]').props().type).toBe('text')
//         wrapper.find('span[className="switch switchOn"]').simulate('click')
//         expect(wrapper.find('input[name="password"]').props().type).toBe('password')
//         expect(wrapper.find('input[name="rePassword"]').props().type).toBe('password')
//         done()
//     }, 0)
// })

test('validate firstName', done => {
    const props = {
        ...propsFromRedux
    }
    const wrapper = mount (
        <CaregiverRegister {...props} />
    )

    setTimeout(() => {
        wrapper.update()

        // Failure case
        wrapper.find('input[name="firstName"]').simulate('change', {
            target: {
                name: 'firstName',
                value: ''
            }
        })
        wrapper.find('input[name="firstName"]').simulate('blur')

        expect(toJson(wrapper.find('.textError').render())).toMatchSnapshot()

        // Success Case

        wrapper.find('input[name="firstName"]').simulate('change', {
            target: {
                name: 'firstName',
                value: 'firstName'
            }
        })
        wrapper.find('input[name="firstName"]').simulate('blur')

        expect(wrapper.find('.textError').length).toBe(0)

        done()
    }, 0)

})

test('validate lastName', done => {
    const props = {
        ...propsFromRedux
    }
    const wrapper = mount (
        <CaregiverRegister {...props} />
    )

    setTimeout(() => {
        wrapper.update()

        // Failure case
        wrapper.find('input[name="lastName"]').simulate('change', {
            target: {
                name: 'lastName',
                value: ''
            }
        })
        wrapper.find('input[name="lastName"]').simulate('blur')

        expect(toJson(wrapper.find('.textError').render())).toMatchSnapshot()

        // Success Case

        wrapper.find('input[name="lastName"]').simulate('change', {
            target: {
                name: 'lastName',
                value: 'lastName'
            }
        })
        wrapper.find('input[name="lastName"]').simulate('blur')

        expect(wrapper.find('.textError').length).toBe(0)

        done()
    }, 0)

})

test('email format must be validated and errors displayed when email input field has lost focus', done => {
    const props = {
        ...propsFromRedux
    }
    const wrapper = mount (
        <CaregiverRegister {...props} />
    )

    setTimeout(() => {
        wrapper.update()

        // Failure case
        wrapper.find('input[name="email"]').simulate('change', {
            target: {
                name: 'email',
                value: 'test'
            }
        })
        wrapper.find('input[name="email"]').simulate('blur')

        expect(toJson(wrapper.find('.textError').render())).toMatchSnapshot()

        // Success Case

        wrapper.find('input[name="email"]').simulate('change', {
            target: {
                name: 'email',
                value: 'test@test.com'
            }
        })
        wrapper.find('input[name="email"]').simulate('blur')

        expect(wrapper.find('.textError').length).toBe(0)

        done()
    }, 0)

})

test('password validation for 1st password entry box is done after user finished typing', done => {
    const props = {
        ...propsFromRedux
    }
    const wrapper = mount (
        <CaregiverRegister {...props} />
    )
    setTimeout(() => {
        wrapper.update()
        // Failure case

        wrapper.find('input[name="password"]').simulate('change', {
            target: {
                name: 'password',
                value: 'test'
            }
        })
        wrapper.find('input[name="password"]').simulate('blur')

        expect(toJson(wrapper.find('.textError').render())).toMatchSnapshot()

        // Success Case

        wrapper.find('input[name="password"]').simulate('change', {
            target: {
                name: 'password',
                value: 'A12345678*a'
            }
        })
        wrapper.find('input[name="password"]').simulate('blur')

        expect(wrapper.find('.textError').length).toBe(0)
        done()
    }, 0)
})

test('password validation for 2nd password entry box must be done as user is typing', done => {
    const props = {
        ...propsFromRedux
    }
    const wrapper = mount (
        <CaregiverRegister {...props} />
    )

    setTimeout(() => {
        wrapper.update()

        wrapper.find('input[name="password"]').simulate('change', {
            target: {
                name: 'password',
                value: 'A12345678*a'
            }
        })
        wrapper.find('input[name="password"]').simulate('blur')

        // Failure case

        wrapper.find('input[name="rePassword"]').simulate('change', {
            target: {
                name: 'rePassword',
                value: 'test'
            }
        })

        expect(toJson(wrapper.find('.textError').render())).toMatchSnapshot()

        // Success case

        wrapper.find('input[name="rePassword"]').simulate('change', {
            target: {
                name: 'rePassword',
                value: 'A12345678*a'
            }
        })

        expect(wrapper.find('.textError').length).toBe(0)
        done()
    }, 0)
})

test('clicks on next button | register api called | login success | save auth token | go to NOPP pages', (done) => {
    const date = new Date()
    const props = {
        ...propsFromRedux,
        page: {
            ...propsFromRedux.page,
            apiData: {
                ...propsFromRedux.page.apiData,
            },
            error: {
                status: null
            }
        },
        callSyncAPIGateway: (name, get, post) => {
            if (name === 'usersOrganizationPolicyIdGet') {
                return Promise.resolve()
            }
            expect(name).toBe('usersRegisterPost')
            expect(post).toEqual({
                "email": "test@test.com",
                "firstName": "firstName",
                "lastName": "lastName",
                "password": "A12345678*a",
                "role": "user",
                "nopp": {
                    'date': date.getFullYear() + '-' + ('0' + (date.getMonth() + 1)).slice(-2) + '-' + date.getDate(),
                    'vEnd': 'v1.0',
                    'vNopp': 'v1.0',
                    'vTerms': 'v1.0',
                }
            })
            props.page.apiData = {...mockObj.auth}
            return Promise.resolve()
        },
        saveAuth: (auth) => {
            expect(auth).toEqual(mockObj.auth)
            return Promise.resolve()
        },
        apiInited: () => {
            return Promise.resolve()
        },
        push: (url) => {
            expect(url).toBe('/caregiver-patient-lookup')
            done()
        }
    }

    const wrapper = mount (
        <CaregiverRegister {...props} />
    )

    setTimeout(() => {
        wrapper.update()
        wrapper.find('input[name="firstName"]').simulate('change', {
            target: {
                name: 'firstName',
                value: 'firstName'
            }
        })
        wrapper.find('input[name="lastName"]').simulate('change', {
            target: {
                name: 'lastName',
                value: 'lastName'
            }
        })
        wrapper.find('input[name="email"]').simulate('change', {
            target: {
                name: 'email',
                value: 'test@test.com'
            }
        })
        wrapper.find('input[name="password"]').simulate('change', {
            target: {
                name: 'password',
                value: 'A12345678*a'
            }
        })
        wrapper.find('input[name="rePassword"]').simulate('change', {
            target: {
                name: 'rePassword',
                value: 'A12345678*a'
            }
        })
        wrapper.find('button[type="submit"]').simulate('submit')
    }, 0)
})

test('register API returns 409 Conflict, display error and link to Login page', (done) => {
    const props = {
        ...propsFromRedux,
        page: {
            ...propsFromRedux.page,
            apiData: {
                ...propsFromRedux.page.apiData,
            },
            error: {
                status: 409
            }
        },
        callSyncAPIGateway: () => {
            return Promise.resolve()
        }
    }

    const wrapper = mount (
        <CaregiverRegister {...props} />
    )

    setTimeout(() => {
        wrapper.update()
        wrapper.find('input[name="firstName"]').simulate('change', {
            target: {
                name: 'firstName',
                value: 'firstName'
            }
        })
        wrapper.find('input[name="lastName"]').simulate('change', {
            target: {
                name: 'lastName',
                value: 'lastName'
            }
        })
        wrapper.find('input[name="email"]').simulate('change', {
            target: {
                name: 'email',
                value: 'test@test.com'
            }
        })
        wrapper.find('input[name="password"]').simulate('change', {
            target: {
                name: 'password',
                value: 'A12345678*a'
            }
        })
        wrapper.find('input[name="rePassword"]').simulate('change', {
            target: {
                name: 'rePassword',
                value: 'A12345678*a'
            }
        })
        wrapper.find('button[type="submit"]').simulate('submit')
        setTimeout(() => {
            wrapper.update()
            expect(toJson(wrapper.find('.textError').render())).toMatchSnapshot()
            done()
        }, 500)
    }, 0)
})

test('clicks on Notice of Privacy Practices | Policies popup show', done => {
    const props = {
        ...propsFromRedux
    }
    const wrapper = mount (
        <CaregiverRegister {...props} />
    )
    setTimeout(() => {
        wrapper.update()
        wrapper.find('#policies-button').simulate('click')
        expect(wrapper.find('.policyWrapper').length).toBe(1)
        done()
    }, 0)
})

test.skip('The "Create Caregiver Account" is disabled until the checkbox to accept the nopp,terms is selected', done => {
    const props = {
        ...propsFromRedux
    }
    const wrapper = mount (
        <CaregiverRegister {...props} />
    )
    setTimeout(() => {
        wrapper.update()
        expect(toJson(wrapper.find('button').render())).toMatchSnapshot()
        wrapper.find('input[type="checkbox"]').simulate('change')
        wrapper.update()
        expect(toJson(wrapper.find('button').render())).toMatchSnapshot()
        done()
    }, 0)
})
