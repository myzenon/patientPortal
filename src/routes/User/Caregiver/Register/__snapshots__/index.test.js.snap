// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The "Create Caregiver Account" is disabled until the checkbox to accept the nopp,terms is selected 1`] = `
<button
  class="button btn btn-secondary disabled"
  disabled=""
  type="submit"
>
  Create Caregiver Account
</button>
`;

exports[`The "Create Caregiver Account" is disabled until the checkbox to accept the nopp,terms is selected 2`] = `
<button
  class="button btn btn-secondary"
  type="submit"
>
  Create Caregiver Account
</button>
`;

exports[`email format must be validated and errors displayed when email input field has lost focus 1`] = `
<div
  class="textError"
>
  Please enter a valid email address.
</div>
`;

exports[`password validation for 1st password entry box is done after user finished typing 1`] = `
<div
  class="textError"
>
  Password must be at least 8 characters long and contain at least one character, one number and one special characters including !@#$%*=
</div>
`;

exports[`password validation for 2nd password entry box must be done as user is typing 1`] = `
<div
  class="textError"
>
  Your passwords do not match, try again.
</div>
`;

exports[`register API returns 409 Conflict, display error and link to Login page 1`] = `
<div
  class="textError"
>
  This email address is already registered.
  <br />
  Click below to Login.
</div>
`;

exports[`renders correctly 1`] = `
<div
  class="container-fluid"
>
  <div
    class="content container"
  >
    <div
      class="heading2"
    >
      Let's get you set up with a caregiver account .
    </div>
    <div
      class="formWrapper"
    >
      <form
        class=""
      >
        <span>
          <div
            class="row"
          >
            <div
              class="col-12 col-sm-6"
            >
              <input
                class="inputText"
                name="firstName"
                placeholder="First Name"
                type="text"
                value=""
              />
            </div>
            <div
              class="col-12 col-sm-6"
            >
              <input
                class="inputText"
                name="lastName"
                placeholder="Last Name"
                type="text"
                value=""
              />
            </div>
          </div>
        </span>
        <div>
          <input
            class="inputText"
            name="email"
            placeholder="email"
            type="email"
            value=""
          />
        </div>
        <div
          class="passwordwrapper"
        >
          <input
            autocomplete="new-password"
            class="passwordText"
            ispassword="true"
            name="password"
            placeholder="password"
            type="text"
            value=""
          />
          <button
            class="showHideButton btn btn-secondary"
            type="button"
          >
            Hide 
          </button>
        </div>
        <div
          class="passwordwrapper"
        >
          <input
            class="passwordText"
            ispassword="true"
            name="rePassword"
            placeholder="re-enter password"
            type="text"
            value=""
          />
          <button
            class="showHideButton btn btn-secondary"
            type="button"
          >
            Hide 
          </button>
        </div>
        <div
          class="formHint"
        >
          <div>
            Password must:
            <ul>
              <li>
                - 8 characters
              </li>
              <li>
                - One uppercase letter
              </li>
              <li>
                - One lowercase letter
              </li>
              <li>
                - One number
              </li>
              <li>
                - One special character !@#$%^&*=
              </li>
            </ul>
          </div>
        </div>
        <div
          class="checkboxWrapper row"
        >
          <div
            class="formLabel col-12 col-md-12"
          >
            <label
              class="checkboxContainer"
            >
              <input
                id="chkCertifyCaregiverRegister"
                type="checkbox"
                value=""
              />
              <span
                class="checkbox"
              />
            </label>
                     I certify that all info I provide is correct and I have read and agree to the  
            <span
              class="noppCheckBoxWrapper"
            >
              <span
                id="policies-button"
              >
                Notice of Privacy Practices, Terms of Service, End User License Agreement and Access to Another Adult's Online Medical Record
              </span>
            </span>
          </div>
        </div>
        <div>
          <button
            class="button btn btn-secondary disabled"
            disabled=""
            id="btnCreateCaregiver"
            type="submit"
          >
            Create Caregiver Account
          </button>
        </div>
      </form>
    </div>
  </div>
  <div
    class="footerWrapper container"
  >
    <div>
      <div
        class="wrapper"
      >
        <div
          class="stepActive"
        >
          Account
          <br />
          Setup
          <div
            class="stepNumberActive"
          >
            1
          </div>
        </div>
        <div
          class="step"
        >
          Medication
          <br />
          Review
          <div
            class="stepNumber"
          >
            2
          </div>
        </div>
        <div
          class="step"
        >
          Additional 
          <br />
          Medications
          <div
            class="stepNumber"
          >
            3
          </div>
        </div>
        <div
          class="step"
        >
          Health
          <br />
          Updates
          <div
            class="stepNumber"
          >
            4
          </div>
        </div>
        <div
          class="step"
        >
          Set
          <br />
          Appointment
          <div
            class="stepNumber"
          >
            5
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`validate firstName 1`] = `
<div
  class="textError"
>
  Please enter your first name.
</div>
`;

exports[`validate lastName 1`] = `
<div
  class="textError"
>
  Please enter your last name.
</div>
`;
